src/Middleware/RequestThrottling/test/SemaphoreWrapperTests.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Xunit;
+using System.Threading;
+using System.Threading.Tasks;
+using System;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Testing;
+
+namespace Microsoft.AspNetCore.RequestThrottling.Tests
+{
+    public class SemaphoreWrapperTests
+    {
+        [Fact]
+        public async Task TracksQueueLength()
+        {
+            using var s = new SemaphoreWrapper(1);
+            Assert.Equal(1, s.Count);
+
+            await s.EnterQueue();
+            Assert.Equal(0, s.Count);
+
+            s.LeaveQueue();
+            Assert.Equal(1, s.Count);
+        }
+
+        [Fact]
+        public void DoesNotWaitIfSpaceAvailible()
+        {
+            using var s = new SemaphoreWrapper(2);
+
+            var t1 = s.EnterQueue();
+            Assert.True(t1.IsCompleted);
+
+            var t2 = s.EnterQueue();
+            Assert.True(t2.IsCompleted);
+
+            var t3 = s.EnterQueue();
+            Assert.False(t3.IsCompleted);
+        }
+
+        [Fact]
+        public async Task WaitsIfNoSpaceAvailible()
+        {
+            using var s = new SemaphoreWrapper(1);
+            await s.EnterQueue();
+
+            var waitingTask = s.EnterQueue();
+            Assert.False(waitingTask.IsCompleted);
+
+            s.LeaveQueue();
+            await waitingTask.TimeoutAfter(TimeSpan.FromSeconds(30));
+        }
+
+        [Fact]
+        public async Task IsEncapsulated()
+        {
+            using var s1 = new SemaphoreWrapper(1);
+            using var s2 = new SemaphoreWrapper(1);
+
+            await s1.EnterQueue();
```

Comment:
Everything you await in a test should have a timeout. I would create an `OrTimeout()` extension method that defaults to 30 seconds.

Reply:
Oh yeah I thought since I can assume certain conditions based on past tests (ie that something can enter) that I wouldn't need to validate that again. But the timeouts aren't about validation or logical consistency, they are about automated tests actually running to completion.

Reply:
The timeouts are basically there just in case you get something wrong, so that the test doesn't just hang indefinitely, but fails at some point :).

We have code in a bunch of places to do this (Kestrel, SignalR, etc.). I wonder if we should look at centralizing that in `src/Shared`. For now just copy the one from SignalR I think.